Split LLVM intrinsic abi handling from the rest of the abi handling

bjorn3 opened the pull request:
LLVM intrinsics have weird requirements like requiring the fake "unadjusted" abi, not being callable through function pointers and for all codegen backends other than cg_llvm requiring special cases to redirect them to the correct backend specific intrinsic (or directly codegen their implementation inline without any intrinsic call). By splitting the LLVM intrinsic handling it becomes easier for backends to special case them and should in the future allow getting rid of the abi calculation for `extern "unadjusted"` in favor of computing the correct abi directly in the backend without depending on the exact way cg_ssa lowers types.

bjorn3 committed:
Split out ABI handling for LLVM intrinsics

bjorn3 committed:
Move llvm intrinsic call to backend

bjorn3 commented:
cc @sayantn This should make https://github.com/rust-lang/rust/pull/140763 much less intrusive for the non-intrinsic call handling and should allow you to localize all changes to cg_llvm without having to touch cg_ssa and cg_gcc.

sayantn commented:
yes, I did `cargo build -p core_arch --tests`, which will compile all intrinsics in x86 (due to all intrinsics having corresponding tests), I am not so sure about aarch64, there can be some intrinsics that are not tested.

bjorn3 committed:
Partially inline get_fn_addr/get_fn in codegen_llvm_intrinsic_call

This moves all LLVM intrinsic handling out of the regular call path for
cg_gcc and makes it easier to hook into this code for future cg_llvm
changes.

bjorn3 committed:
Don't assume get_fn is only called from codegen_mir in cg_gcc

bjorn3 committed:
Pass Function to Builder::function_call in cg_gcc

bjorn3 committed:
Inline BuilderMethods::call for intrinsics

Intrinsics only need a fraction of the functionality offered by
BuilderMethods::call and in particular don't need the FnAbi to be
computed other than (currently) as step towards computing the function
value type.

antoyo reviewed (approved):
The `stdarch` tests pass with this branch.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/148533/c8f22ca269a1f2653ac962fe2bc21105065fd6cd..6d4eb0e3f3137cabed500e89e6c7d50ff4a117d8)*

bjorn3 committed:
Fix for uninhabited llvm intrinsic return type

purplesyringa commented:
That's probably a good idea, I'll see if I can make a PR for that in a bit. (Just to be clear, `llvm.wasm.rethrow` is also unwinding, we just never had a need to use it.)

bjorn3 committed:
Add a hack for llvm.wasm.throw

As this is the only unwinding intrinsic we use and
codegen_llvm_intrinsic_call currently doesn't handle unwinding
intrinsics, this uses the conventional call path for llvm.wasm.throw.

sayantn commented on compiler/rustc_codegen_llvm/src/intrinsic.rs:
```diff
@@ -609,6 +614,99 @@ impl<'ll, 'tcx> IntrinsicCallBuilderMethods<'tcx> for Builder<'_, 'll, 'tcx> {
+            let sym = tcx.symbol_name(instance).name;
+
+            // FIXME use get_intrinsic
+            let llfn = if let Some(llfn) = self.get_declared_value(sym) {
```
you don't really need this branch, LLVM already does this in `llvm::getOrInsertFunction`

bors commented:
:umbrella: The latest upstream changes (presumably #149351) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:pushpin: Commit 4eb6e1c1a33f177226090097493a8b0c5897eed9 has been approved by `WaffleLapkin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=WaffleLapkin 4eb6e1c1a33f177226090097493a8b0c5897eed9 -->
<!-- homu: {"type":"Approved","sha":"4eb6e1c1a33f177226090097493a8b0c5897eed9","approver":"WaffleLapkin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bjorn3 commented:
I can't reproduce this locally after a rebase.

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/20529434578/job/58978504097) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/58978504097) [(plain)](https://github.com/rust-lang/rust/commit/7712c1aafd0a6fb5c1dc7a5bc494a39fa2eaba10/checks/58978504097/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    |
    = help: items from traits can only be used if the trait is in scope
help: one of the expressions' fields has a method of the same name
    |
627 |                 let fn_abi = self.cx.tcx.fn_abi_of_instance(instance, ty::List::empty());
    |                                   +++++++
help: one of the expressions' fields has a method of the same name
    |
627 |                 let fn_abi = self.tcx.fn_abi_of_instance(instance, ty::List::empty());
    |                                   ++++
help: trait `FnAbiOf` which provides `fn_abi_of_instance` is implemented but not in scope; perhaps you want to import it
    |
  1 + use crate::rustc_middle::ty::layout::FnAbiOf;
    |
help: there is a method `fn_abi_of_fn_ptr` with a similar name
    |
```

</details>

bjorn3 committed:
Fix compilation of cg_gcc with master feature disabled

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/20536893225/job/58996316295) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/58996316295) [(plain)](https://github.com/rust-lang/rust/commit/813fa6ca7a062d0338d9ccd838aa3511ad1765ab/checks/58996316295/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] gccjit_sys test:false 0.251
    Checking gccjit v3.1.1
[RUSTC-TIMING] gccjit test:false 0.574
    Checking rustc_codegen_gcc v0.1.0 (/checkout/compiler/rustc_codegen_gcc)
error: internal compiler error: compiler/rustc_middle/src/ty/context.rs:2986:17: args not compatible with generics for abi::FnAbiGccExt: args=[
                                    &'?65 rustc_target::callconv::FnAbi<'?66, rustc_middle::ty::Ty<'tcx/#2>>,
                                ], generics=[
                                    Self/#0,
                                    'gcc/#1,
                                    'tcx/#2,
                                ]


thread 'rustc' (36278) panicked at compiler/rustc_middle/src/ty/context.rs:2986:17:
Box<dyn Any>
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: <rustc_errors::diagnostic::BugAbort as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
   2: rustc_middle::util::bug::opt_span_bug_fmt::<rustc_span::span_encoding::Span>::{closure#0}
   3: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}
   4: rustc_middle::ty::context::tls::with_context_opt::<rustc_middle::ty::context::tls::with_opt<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
   5: rustc_middle::util::bug::bug_fmt
   6: <rustc_middle::ty::context::TyCtxt>::debug_assert_args_compatible
   7: <rustc_hir_typeck::fn_ctxt::FnCtxt>::detect_and_explain_multiple_crate_versions_of_trait_item
   8: <rustc_hir_typeck::fn_ctxt::FnCtxt>::suggest_traits_to_import
   9: <rustc_hir_typeck::fn_ctxt::FnCtxt>::report_no_match_method_error
  10: <rustc_hir_typeck::fn_ctxt::FnCtxt>::report_method_error
  11: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  12: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  13: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation
  14: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_decl
  15: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
  16: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  17: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  18: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation
  19: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  20: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  21: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation
  22: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_decl
  23: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
  24: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  25: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  26: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation
  27: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  28: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  29: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation
  30: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_decl
  31: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
  32: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  33: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  34: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation
  35: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_return_or_body_tail
  36: rustc_hir_typeck::check::check_fn
  37: rustc_hir_typeck::typeck_with_inspect
      [... omitted 2 frames ...]
  38: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis::check_crate::{closure#2}>::{closure#0}
  39: rustc_data_structures::sync::parallel::par_for_each_in::<&rustc_span::def_id::LocalDefId, &[rustc_span::def_id::LocalDefId], <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners<rustc_hir_analysis::check_crate::{closure#2}>::{closure#0}>
  40: rustc_hir_analysis::check_crate
  41: rustc_interface::passes::analysis
      [... omitted 2 frames ...]
  42: <std::thread::local::LocalKey<core::cell::Cell<*const ()>>>::with::<rustc_middle::ty::context::tls::enter_context<<rustc_middle::ty::context::GlobalCtxt>::enter<rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>::{closure#1}, core::option::Option<rustc_interface::queries::Linker>>::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>
  43: <rustc_middle::ty::context::TyCtxt>::create_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}>
  44: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  45: <alloc::boxed::Box<dyn for<'a> core::ops::function::FnOnce<(&'a rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &'a std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena<'a>>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}), Output = core::option::Option<rustc_interface::queries::Linker>>> as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once
  46: rustc_interface::passes::create_and_enter_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>
  47: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>
  48: rustc_span::create_session_globals_then::<(), rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---
warning: the ICE couldn't be written to `/checkout/rustc-ice-2025-12-27T09_17_48-36277.txt`: Read-only file system (os error 30)

note: rustc 1.94.0-nightly (977fbd785 2025-12-27) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type dylib -C opt-level=3 -C embed-bitcode=no -C debug-assertions=on -C symbol-mangling-version=v0 -Z annotate-moves -Z unstable-options -Z force-unstable-if-unmarked -Z macro-backtrace -C split-debuginfo=off -C link-arg=-L/checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm/lib -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -Z unstable-options -Z binary-dep-depinfo -Z on-broken-pipe=kill -Z force-unstable-if-unmarked

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [typeck] type-checking `intrinsic::<impl at compiler/rustc_codegen_gcc/src/intrinsic/mod.rs:288:1: 288:83>::codegen_llvm_intrinsic_call`
#1 [analysis] running analysis passes on crate `rustc_codegen_gcc`
end of query stack
note: Clippy version: clippy 0.1.94 (977fbd7858 2025-12-27)

[RUSTC-TIMING] rustc_codegen_gcc test:false 1.329
error: could not compile `rustc_codegen_gcc` (lib)

Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc /checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/clippy-driver /checkout/obj/build/bootstrap/debug/rustc --crate-name rustc_codegen_gcc --edition=2024 compiler/rustc_codegen_gcc/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type dylib --emit=dep-info,metadata -C opt-level=3 -C embed-bitcode=no -C debug-assertions=on --check-cfg 'cfg(docsrs,test)' --check-cfg 'cfg(feature, values("default", "master"))' -C metadata=6ee6b087753e33e5 -C extra-filename=-d969d74493bb4e34 --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage2-codegen/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-codegen/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-codegen/release/deps --extern gccjit=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-codegen/x86_64-unknown-linux-gnu/release/deps/libgccjit-9f3810b249026cb3.rmeta --extern object=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-codegen/x86_64-unknown-linux-gnu/release/deps/libobject-fd0c5b04dfd81ef0.rmeta --extern tempfile=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-codegen/x86_64-unknown-linux-gnu/release/deps/libtempfile-e4b66f881fa3bc88.rmeta --sysroot /checkout/obj/build/x86_64-unknown-linux-gnu/stage1 -Csymbol-mangling-version=v0 -Zannotate-moves -Zunstable-options '--check-cfg=cfg(bootstrap)' -Zforce-unstable-if-unmarked -Zmacro-backtrace -Csplit-debuginfo=off -Clink-arg=-L/checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm/lib -Clink-args=-Wl,-z,origin '-Clink-args=-Wl,-rpath,$ORIGIN/../lib' -Alinker-messages -Zunstable-options -Z binary-dep-depinfo` (exit status: 101)
Bootstrap failed while executing `clippy ci --stage 2`
Build completed unsuccessfully in 0:11:36
  local time: Sat Dec 27 09:17:50 UTC 2025
  network time: Sat, 27 Dec 2025 09:17:50 GMT
##[error]Process completed with exit code 1.
```

</details>

GuillaumeGomez commented:
Add `--no-default-features`, it will not build with the `master` feature.

GuillaumeGomez commented:
Sorry misunderstood what you asked. You can't do that currently unless when running `x.py clippy cg_gcc` (where you don't need to pass any flag, it checks both with and without the `master` feature). I'm not sure it would make sense to add such an option. You can add `cargo.arg("--no-default-features");` [here](https://github.com/rust-lang/rust/blob/99cc3799b5b3cabdb7bb233f43390474103c98e0/src/bootstrap/src/core/build_steps/compile.rs#L1721) to use this flag as a temporary way to be able to check it.

bjorn3 commented:
I can reproduce the ICE with `./x.py clippy rustc_codegen_gcc --stage 2`.

bors commented:
:pushpin: Commit aeac6cb2d65499eb4eb541b66dbf292ffe93d770 has been approved by `WaffleLapkin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=WaffleLapkin aeac6cb2d65499eb4eb541b66dbf292ffe93d770 -->
<!-- homu: {"type":"Approved","sha":"aeac6cb2d65499eb4eb541b66dbf292ffe93d770","approver":"WaffleLapkin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
